<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Adjusted p-values for simple multiple testing procedures — mt.rawp2adjp • switchgrassGWAS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Adjusted p-values for simple multiple testing procedures — mt.rawp2adjp"><meta property="og:description" content="This function computes adjusted p-values for simple multiple
testing procedures from a vector of raw (unadjusted) p-values. The
procedures include the Bonferroni, Holm (1979), Hochberg (1988), and
Sidak procedures for strong control of the family-wise Type I error
rate (FWER), and the Benjamini &amp;amp; Hochberg (1995) and Benjamini &amp;amp;
Yekutieli (2001) procedures for (strong) control of the false discovery
rate (FDR). The less conservative adaptive Benjamini &amp;amp; Hochberg (2000)
and two-stage Benjamini &amp;amp; Hochberg (2006) FDR-controlling procedures are
also included. This function is taken from the multtest package. It is
the only function used from this package and is added to this package
wholesale to reduce user installation burden."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/fbm.html">Using file-backed matrices - FBM</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/inputs.html">SNP file inputs and QC</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Adjusted p-values for simple multiple testing procedures</h1>
    
    <div class="hidden name"><code>mt.rawp2adjp.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function computes adjusted p-values for simple multiple
testing procedures from a vector of raw (unadjusted) p-values. The
procedures include the Bonferroni, Holm (1979), Hochberg (1988), and
Sidak procedures for strong control of the family-wise Type I error
rate (FWER), and the Benjamini &amp; Hochberg (1995) and Benjamini &amp;
Yekutieli (2001) procedures for (strong) control of the false discovery
rate (FDR). The less conservative adaptive Benjamini &amp; Hochberg (2000)
and two-stage Benjamini &amp; Hochberg (2006) FDR-controlling procedures are
also included. This function is taken from the multtest package. It is
the only function used from this package and is added to this package
wholesale to reduce user installation burden.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">mt.rawp2adjp</span><span class="op">(</span><span class="va">rawp</span>, proc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Bonferroni"</span>, <span class="st">"Holm"</span>, <span class="st">"Hochberg"</span>, <span class="st">"SidakSS"</span>,
    <span class="st">"SidakSD"</span>, <span class="st">"BH"</span>, <span class="st">"BY"</span>,<span class="st">"ABH"</span>,<span class="st">"TSBH"</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>rawp</dt>
<dd><p>A vector of raw (unadjusted) p-values for each hypothesis under
consideration. These could be nominal p-values, for example, from
t-tables, or permutation p-values as given in mt.maxT and mt.minP. If
the mt.maxT or mt.minP functions are used, raw p-values should be given
in the original data order, ordered by the index of that data.</p></dd>
<dt>proc</dt>
<dd><p>A vector of character strings containing the names of the
multiple testing procedures for which adjusted p-values are to be
computed. This vector should include any of the following: "Bonferroni",
"Holm", "Hochberg", "SidakSS", "SidakSD", "BH", "BY", "ABH", "TSBH".</p></dd>
<dt>alpha</dt>
<dd><p>A nominal type I error rate, or a vector of error rates, used
for estimating the number of true null hypotheses in the two-stage
Benjamini &amp; Hochberg procedure ("TSBH"). Default is 0.05.</p></dd>
<dt>na.rm</dt>
<dd><p>An option for handling NA values in a list of raw p-values.
If FALSE, the number of hypotheses considered is the length of the vector
of raw p-values. Otherwise, if TRUE, the number of hypotheses is the
number of raw p-values which were not NAs.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with components: adjp, index, h0.ABH, h0.TSBH. See multtest
package on Bioconductor for details.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Sandrine Dudoit, http://www.stat.berkeley.edu/~sandrine,</p>
<p>Yongchao Ge, yongchao.ge@mssm.edu,</p>
<p>Houston Gilbert, http://www.stat.berkeley.edu/~houston.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

