<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Juenger lab standard GWAS function. — pvdiv_standard_gwas • switchgrassGWAS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Juenger lab standard GWAS function. — pvdiv_standard_gwas"><meta property="og:description" content="This function is a wrapper around the standard GWAS procedures
in the Juenger lab. Singular value decomposition of the SNPs is done to
get principal components for population structure correction; the 'best'
number of PCs is chosen as the one that makes lambda_GC, the Genomic
Control coefficient, closest to 1. (See the lambdagc parameter to set
this yourself.) Next, genome-wide association is conducted, and the GWAS
output can be saved, as well as Manhattan plots, QQ-plots, and annotation
information for the top SNPs for each phenotype."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/fbm.html">Using file-backed matrices - FBM</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/inputs.html">SNP file inputs and QC</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Juenger lab standard GWAS function.</h1>
    
    <div class="hidden name"><code>pvdiv_standard_gwas.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is a wrapper around the standard GWAS procedures
in the Juenger lab. Singular value decomposition of the SNPs is done to
get principal components for population structure correction; the 'best'
number of PCs is chosen as the one that makes lambda_GC, the Genomic
Control coefficient, closest to 1. (See the <code>lambdagc</code> parameter to set
this yourself.) Next, genome-wide association is conducted, and the GWAS
output can be saved, as well as Manhattan plots, QQ-plots, and annotation
information for the top SNPs for each phenotype.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">pvdiv_standard_gwas</span><span class="op">(</span>
  <span class="va">snp</span>,
  df <span class="op">=</span> <span class="fu">switchgrassGWAS</span><span class="fu">::</span><span class="va"><a href="pvdiv_phenotypes.html">pvdiv_phenotypes</a></span>,
  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"linear"</span>, <span class="st">"logistic"</span><span class="op">)</span>,
  ncores <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span>,
  outputdir <span class="op">=</span> <span class="st">"."</span>,
  covar <span class="op">=</span> <span class="cn">NULL</span>,
  lambdagc <span class="op">=</span> <span class="cn">TRUE</span>,
  savegwas <span class="op">=</span> <span class="cn">FALSE</span>,
  savetype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rds"</span>, <span class="st">"fbm"</span>, <span class="st">"both"</span><span class="op">)</span>,
  suffix <span class="op">=</span> <span class="st">""</span>,
  saveplots <span class="op">=</span> <span class="cn">TRUE</span>,
  saveannos <span class="op">=</span> <span class="cn">FALSE</span>,
  txdb <span class="op">=</span> <span class="cn">NULL</span>,
  minphe <span class="op">=</span> <span class="fl">200</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>snp</dt>
<dd><p>A "bigSNP" object; load with <code><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">bigsnpr::snp_attach()</a></code>.
Here, genomic information for Panicum virgatum. SNP data
is available at doi:10.18738/T8/ET9UAU</p></dd>
<dt>df</dt>
<dd><p>Dataframe of phenotypes where the first column is PLANT_ID.</p></dd>
<dt>type</dt>
<dd><p>Character string. Type of univarate regression to run for GWAS.
Options are "linear" or "logistic".</p></dd>
<dt>ncores</dt>
<dd><p>Number of cores to use. Default is one.</p></dd>
<dt>outputdir</dt>
<dd><p>String or file.path() to the output directory. Default is
the working directory.</p></dd>
<dt>covar</dt>
<dd><p>Optional covariance matrix to include in the regression. You
can generate these using <code><a href="pvdiv_autoSVD.html">pvdiv_autoSVD()</a></code>.</p></dd>
<dt>lambdagc</dt>
<dd><p>Default is TRUE - should lambda_GC be used to find the best
population structure correction? Alternatively, you can provide a data
frame containing "NumPCs" and the phenotype names containing lambda_GC
values. This is saved to the output directory by pvdiv_standard_gwas and
saved or generated by pvdiv_lambda_GC.</p></dd>
<dt>savegwas</dt>
<dd><p>Logical. Should the gwas output be saved to the
working directory? These files are typically quite large. Default is
FALSE.</p></dd>
<dt>savetype</dt>
<dd><p>Character string. Type of GWAS save file. Options are 'rds',
which saves individual rds files for each GWAS; 'fbm', which saves one
filebacked big matrix (using the bigsnpr package), or 'both', which
saves both file types. These files are typically quite large.</p></dd>
<dt>suffix</dt>
<dd><p>Optional character vector to give saved files a unique search
string/name.</p></dd>
<dt>saveplots</dt>
<dd><p>Logical. Should Manhattan and QQ-plots be generated and
saved to the working directory? Default is TRUE.</p></dd>
<dt>saveannos</dt>
<dd><p>Logical. Should annotation tables for top SNPs be generated
and saved to the working directory? Default is FALSE. Can take
additional arguments; requires a txdb.sqlite object used in
AnnotationDbi.</p></dd>
<dt>txdb</dt>
<dd><p>A txdb object such as 'Pvirgatum_516_v5.1.gene.txdb.sqlite'.
Load this into your environment with AnnotationDbi::loadDb.</p></dd>
<dt>minphe</dt>
<dd><p>Integer. What's the minimum number of phenotyped individuals
to conduct a GWAS on? Default is 200. Use lower values with caution.</p></dd>
<dt>...</dt>
<dd><p>Other arguments to <code><a href="pvdiv_lambda_GC.html">pvdiv_lambda_GC</a></code> or
<code><a href="pvdiv_table_topsnps.html">pvdiv_table_topsnps</a></code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A big_SVD object.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="co"># Here we specify that we do want to generate and save the gwas dataframes,</span></span>
<span class="r-in"><span class="co"># the Manhattan and QQ-plots, and the annotation tables.</span></span>
<span class="r-in"><span class="fu">pvdiv_standard_gwas</span><span class="op">(</span><span class="va">snp</span>, df <span class="op">=</span> <span class="va">pvdiv_phenotypes</span>, type <span class="op">=</span> <span class="st">"linear"</span>, covar <span class="op">=</span> <span class="va">svd</span>,</span>
<span class="r-in">    ncores <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span>, lambdagc <span class="op">=</span> <span class="cn">TRUE</span>, savegwas <span class="op">=</span> <span class="cn">TRUE</span>, saveplots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">    saveannos <span class="op">=</span> <span class="cn">TRUE</span>, txdb <span class="op">=</span> <span class="va">txdb</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"><span class="co"># In this example, we run GWAS on all the phenotypes in pvdiv_phenotypes</span></span>
<span class="r-in"><span class="co"># using an example SNP set of ~1800 SNPs.</span></span>
<span class="r-in"><span class="va">snpfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_bigsnp.rds"</span>, package <span class="op">=</span> <span class="st">"switchgrassGWAS"</span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span>
<span class="r-in"><span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="va">snpfile</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">pvdiv_standard_gwas</span><span class="op">(</span><span class="va">snp</span>, df <span class="op">=</span> <span class="va">pvdiv_phenotypes</span>, type <span class="op">=</span> <span class="st">"linear"</span>, savegwas <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span class="r-in">    saveplots <span class="op">=</span> <span class="cn">FALSE</span>, ncores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'lambdagc' is TRUE, so lambda_GC will be used to find the best population structure correction using the covariance matrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'savegwas' is FALSE, so the gwas results will not be saved to disk.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Covariance matrix (covar) was not supplied - this will be generated using pvdiv_autoSVD().</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'saveoutput' is FALSE, so the svd will not be saved to the working directory.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for GWAS_CT.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for GWAS_CT using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: GWAS_CT</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for MAT.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for MAT using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: MAT</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for bio17.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio17 using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: bio17</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for bio4.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio4 using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: bio4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for bio16.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio16 using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: bio16</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for AHM.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for AHM using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: AHM</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for bio2.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio2 using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: bio2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for bio5.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for bio5 using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: bio5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for FRAC_SRV_THREE.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for FRAC_SRV_THREE using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: FRAC_SRV_THREE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for CLMB_BIOMASS.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for CLMB_BIOMASS using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: CLMB_BIOMASS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for KBSM_BIOMASS.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for KBSM_BIOMASS using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: KBSM_BIOMASS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now starting GWAS pipeline for PKLE_BIOMASS.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 0 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 1 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 2 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 3 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 4 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 5 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 6 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 7 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 8 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 9 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 10 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 11 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 12 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 13 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 14 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished Lambda_GC calculation for PKLE_BIOMASS using 15 PCs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished phenotype 1: PKLE_BIOMASS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Now running GWAS with the best population structure correction.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="r-in"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

