<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Annotation tables for top SNPs • switchgrassGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Annotation tables for top SNPs">
<meta property="og:description" content="switchgrassGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="annotations_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Annotation tables for top SNPs</h1>
                        <h4 class="author">Alice MacQueen</h4>
            
            <h4 class="date">2020-01-07</h4>
      
      
      <div class="hidden name"><code>annotations.Rmd</code></div>

    </div>

    
    
<div id="annotations-for-top-association" class="section level2">
<h2 class="hasAnchor">
<a href="#annotations-for-top-association" class="anchor"></a>Annotations for Top Association</h2>
<div id="load-annotation-information-for-switchgrass" class="section level3">
<h3 class="hasAnchor">
<a href="#load-annotation-information-for-switchgrass" class="anchor"></a>Load annotation information for switchgrass</h3>
<p>The function <code>pvdiv_table_topsnps</code> creates dataframes containing annotation information for the switchgrass diversity panel. To use this function, first load the provided annotation information from wherever you have saved it (currently, this is version 5.1 of the annotation information for <em>Panicum virgatum</em>):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi">AnnotationDbi</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">loadDb</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"Pvirgatum_516_v5.1.gene.txdb.sqlite"</span><span class="op">)</span></code></pre></div>
</div>
<div id="load-snp-dataset-and-gwas-results" class="section level3">
<h3 class="hasAnchor">
<a href="#load-snp-dataset-and-gwas-results" class="anchor"></a>Load SNP dataset and GWAS results</h3>
<p>We will use the same toy SNP dataset that we used for <a href="https://alice-macqueen.github.io/switchgrassGWAS/articles/gwas.html">genome-wide association</a>, and find annotations for the same phenotype, GWAS_CT.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/">bigsnpr</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: bigstatsr</span>

<span class="va">snpfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_bigsnp.rds"</span>, package <span class="op">=</span> <span class="st">"switchgrassGWAS"</span><span class="op">)</span>
<span class="va">gwasfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"GWAS_datatable_example_GWAS_CT.rds"</span>, 
                        package <span class="op">=</span> <span class="st">"switchgrassGWAS"</span><span class="op">)</span>
                        
<span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">gwasfile</span><span class="op">)</span>
<span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="va">snpfile</span><span class="op">)</span></code></pre></div>
<p>Then, specify the number of top SNPs you want to find annotation information for, a FDR threshold to find annotation information for, and a set of distances away from the associated SNP (in bp) for which to pull annotations. Here, we find genes 1kb and 20kb away from 1) top 10 associations and 2) associations above a FDR of 10%.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">anno_tables</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pvdiv_table_topsnps.html">pvdiv_table_topsnps</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">gwas</span>, type <span class="op">=</span> <span class="st">"bigsnp"</span>, n <span class="op">=</span> <span class="fl">10</span>, 
                                  FDRalpha <span class="op">=</span> <span class="fl">0.1</span>, rangevector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">20000</span><span class="op">)</span>,
                                  snp <span class="op">=</span> <span class="va">snp</span>, txdb <span class="op">=</span> <span class="va">txdb</span><span class="op">)</span></code></pre></div>
<p>Let’s look at the annotations for genes within 1000bp of SNPs above the 10% FDR cutoff. In our toy example, there are seven of these:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">kable</span><span class="op">(</span><span class="va">anno_tables</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="3%">
<col width="1%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="3%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="3%">
<col width="3%">
<col width="1%">
<col width="4%">
<col width="0%">
<col width="0%">
<col width="0%">
<col width="0%">
<col width="5%">
<col width="4%">
<col width="5%">
<col width="4%">
<col width="2%">
<col width="11%">
<col width="1%">
<col width="1%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="left">CHR</th>
<th align="right">region_start</th>
<th align="right">region_end</th>
<th align="right">width</th>
<th align="left">region_strand</th>
<th align="left">Annotation</th>
<th align="right">QUERYID</th>
<th align="right">TXID</th>
<th align="left">Gene ID</th>
<th align="right">POS</th>
<th align="right">SNP Effect</th>
<th align="right">SNP standard error</th>
<th align="right">bigsnpscore</th>
<th align="right">p value</th>
<th align="right">log10p</th>
<th align="right">FDR Adjusted p value</th>
<th align="right">gene_start</th>
<th align="right">gene_end</th>
<th align="right">gene_width</th>
<th align="left">gene_strand</th>
<th align="right">pacId</th>
<th align="left">transcriptName</th>
<th align="left">peptideName</th>
<th align="left">Pfam</th>
<th align="left">Panther Categories</th>
<th align="left">KOG</th>
<th align="left">ec</th>
<th align="left">KO</th>
<th align="left">GO</th>
<th align="left">Arabidopsis thaliana homolog</th>
<th align="left">A. thaliana gene name</th>
<th align="left">A. thaliana gene annotation</th>
<th align="left">Rice homolog</th>
<th align="left">Rice gene name</th>
<th align="left">Rice gene annotation</th>
<th align="right">d2_start</th>
<th align="right">d2_end</th>
<th align="right">distance from gene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Chr01N</td>
<td align="right">34181686</td>
<td align="right">34182686</td>
<td align="right">1001</td>
<td align="left">*</td>
<td align="left">intergenic</td>
<td align="right">1</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">34182186</td>
<td align="right">0.9888194</td>
<td align="right">0.2562594</td>
<td align="right">3.858665</td>
<td align="right">0.0001261</td>
<td align="right">3.899256</td>
<td align="right">0.0756650</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">Chr02K</td>
<td align="right">68243567</td>
<td align="right">68244567</td>
<td align="right">1001</td>
<td align="left">*</td>
<td align="left">intergenic</td>
<td align="right">2</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">68244067</td>
<td align="right">-0.9049649</td>
<td align="right">0.2237918</td>
<td align="right">-4.043781</td>
<td align="right">0.0000593</td>
<td align="right">4.226830</td>
<td align="right">0.0533842</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">Chr03N</td>
<td align="right">31494893</td>
<td align="right">31495893</td>
<td align="right">1001</td>
<td align="left">*</td>
<td align="left">intergenic</td>
<td align="right">3</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">31495393</td>
<td align="right">1.2009167</td>
<td align="right">0.2758415</td>
<td align="right">4.353647</td>
<td align="right">0.0000157</td>
<td align="right">4.804299</td>
<td align="right">0.0209074</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">Chr05K</td>
<td align="right">49122041</td>
<td align="right">49123041</td>
<td align="right">1001</td>
<td align="left">-</td>
<td align="left">promoter</td>
<td align="right">4</td>
<td align="right">66970</td>
<td align="left">Pavir.5KG551800</td>
<td align="right">49122541</td>
<td align="right">-1.1140440</td>
<td align="right">0.2251390</td>
<td align="right">-4.948250</td>
<td align="right">0.0000010</td>
<td align="right">6.013622</td>
<td align="right">0.0034888</td>
<td align="right">49122303</td>
<td align="right">49122857</td>
<td align="right">555</td>
<td align="left">-</td>
<td align="right">41544150</td>
<td align="left">Pavir.5KG551800.1</td>
<td align="left">Pavir.5KG551800.1.p</td>
<td align="left">PF01167</td>
<td align="left">PTHR16517,PTHR16517:SF32</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT1G53320.1</td>
<td align="left">AtTLP7,TLP7</td>
<td align="left">tubby like protein 7</td>
<td align="left">LOC_Os05g43850.1</td>
<td align="left">NA</td>
<td align="left">OsFBT9 - F-box and tubby domain containing protein, expressed</td>
<td align="right">238</td>
<td align="right">-316</td>
<td align="right">238</td>
</tr>
<tr class="odd">
<td align="left">Chr07N</td>
<td align="right">34540063</td>
<td align="right">34541063</td>
<td align="right">1001</td>
<td align="left">-</td>
<td align="left">promoter</td>
<td align="right">5</td>
<td align="right">96920</td>
<td align="left">Pavir.7NG229717</td>
<td align="right">34540563</td>
<td align="right">-0.7609417</td>
<td align="right">0.1757385</td>
<td align="right">-4.329965</td>
<td align="right">0.0000174</td>
<td align="right">4.758882</td>
<td align="right">0.0209074</td>
<td align="right">34539192</td>
<td align="right">34539791</td>
<td align="right">600</td>
<td align="left">-</td>
<td align="right">41627406</td>
<td align="left">Pavir.7NG229717.1</td>
<td align="left">Pavir.7NG229717.1.p</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">ChrUn.fgenesh.mRNA.22</td>
<td align="left">NA</td>
<td align="left">expressed protein</td>
<td align="right">1371</td>
<td align="right">772</td>
<td align="right">772</td>
</tr>
<tr class="even">
<td align="left">Chr07N</td>
<td align="right">34540063</td>
<td align="right">34541063</td>
<td align="right">1001</td>
<td align="left">*</td>
<td align="left">intergenic</td>
<td align="right">5</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">34540563</td>
<td align="right">-0.7609417</td>
<td align="right">0.1757385</td>
<td align="right">-4.329965</td>
<td align="right">0.0000174</td>
<td align="right">4.758882</td>
<td align="right">0.0209074</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">Chr08K</td>
<td align="right">4378182</td>
<td align="right">4379182</td>
<td align="right">1001</td>
<td align="left">*</td>
<td align="left">intergenic</td>
<td align="right">6</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">4378682</td>
<td align="right">-0.6237182</td>
<td align="right">0.1604175</td>
<td align="right">-3.888093</td>
<td align="right">0.0001121</td>
<td align="right">3.950457</td>
<td align="right">0.0756650</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
</tbody>
</table>
<p><code>pvdiv_table_topsnps</code> will return a list of all of the tables you requested, named according to the criteria used to create the table. You can then save these tables individually as csv or any other type of file.</p>
<p>If the resultant tables are small, say, less than 2000 entries apiece, I favor saving these tables as individual sheets in an Excel file using the R package <code>XLConnect</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">XLConnect</span><span class="op">)</span>

<span class="va">wb1</span> <span class="op">&lt;-</span> <span class="fu">loadWorkbook</span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"Annotation_tables_GWAS_CT.xlsx"</span>, 
                    create <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">anno_tables</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">createSheet</span><span class="op">(</span><span class="va">wb1</span>, name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">anno_tables</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span>
  <span class="fu">writeWorksheet</span><span class="op">(</span><span class="va">wb1</span>, <span class="va">anno_tables</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>, sheet <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">anno_tables</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span>
  <span class="op">}</span>
<span class="fu">saveWorkbook</span><span class="op">(</span><span class="va">wb1</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
