<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genome-Wide Association • switchgrassGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Genome-Wide Association">
<meta property="og:description" content="switchgrassGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/fbm.html">Using file-backed matrices - FBM</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/inputs.html">SNP file inputs and QC</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genome-Wide Association</h1>
                        <h4 data-toc-skip class="author">Alice MacQueen</h4>
            
            <h4 data-toc-skip class="date">2020-01-06</h4>
      
      
      <div class="hidden name"><code>gwas.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="get-genotype-file">Get genotype file<a class="anchor" aria-label="anchor" href="#get-genotype-file"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the example bedfile from the package switchgrassGWAS</span>
<span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.bed"</span>, package <span class="op">=</span> <span class="st">"switchgrassGWAS"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="set-up-snp-and-phenotype-data-frames-">Set up SNP and phenotype data frames.<a class="anchor" aria-label="anchor" href="#set-up-snp-and-phenotype-data-frames-"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load packages bigsnpr and switchgrassGWAS</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">switchgrassGWAS</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span>
<span class="co">#&gt;   method from   </span>
<span class="co">#&gt;   +.gg   ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'bigsnpr' was built under R version 4.1.2</span>
<span class="co">#&gt; Loading required package: bigstatsr</span>
<span class="co">#&gt; Warning: package 'bigstatsr' was built under R version 4.1.2</span>

<span class="co"># Read from bed/bim/fam to create the new files that bigsnpr uses.</span>
<span class="co"># Let's put them in an temporary directory for this demo.</span>
<span class="va">tmpfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="va">tmpfile</span><span class="op">)</span>
<span class="co">#&gt; [1] "C:\\Users\\Alice\\AppData\\Local\\Temp\\RtmpOGx3b2\\file211c2f056bc8.rds"</span>

<span class="co"># Attach the "bigSNP" object to the R session.</span>
<span class="va">snp_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpfile</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># What does the bigSNP object look like?</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">snp_example</span>, max.level <span class="op">=</span> <span class="fl">2</span>, strict.width <span class="op">=</span> <span class="st">"cut"</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 fields</span>
<span class="co">#&gt;   ..and 26 methods, of which 12 are  possibly relevant:</span>
<span class="co">#&gt;   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span>
<span class="co">#&gt;   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,</span>
<span class="co">#&gt;   ..  save, show#envRefClass, show#FBM</span>
<span class="co">#&gt;  $ fam      :'data.frame':   630 obs. of  6 variables:</span>
<span class="co">#&gt;   ..$ family.ID  : chr [1:630] "Pvirgatum" "Pvirgatum" "Pvirgatum" "Pvirgatum"..</span>
<span class="co">#&gt;   ..$ sample.ID  : chr [1:630] "J181.A" "J250.C" "J251.C" "J352.A" ...</span>
<span class="co">#&gt;   ..$ paternal.ID: int [1:630] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ maternal.ID: int [1:630] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ sex        : int [1:630] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ affection  : int [1:630] -9 -9 -9 -9 -9 -9 -9 -9 -9 -9 ...</span>
<span class="co">#&gt;  $ map      :'data.frame':   3600 obs. of  6 variables:</span>
<span class="co">#&gt;   ..$ chromosome  : chr [1:3600] "Chr01K" "Chr01K" "Chr01K" "Chr01K" ...</span>
<span class="co">#&gt;   ..$ marker.ID   : chr [1:3600] "Chr01K_105542" "Chr01K_798650" "Chr01K_1147"..</span>
<span class="co">#&gt;   ..$ genetic.dist: int [1:3600] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ physical.pos: int [1:3600] 105542 798650 1147074 1719551 1915851 1965986..</span>
<span class="co">#&gt;   ..$ allele1     : chr [1:3600] "A" "C" "T" "T" ...</span>
<span class="co">#&gt;   ..$ allele2     : chr [1:3600] "G" "G" "C" "G" ...</span>
<span class="co">#&gt;  - attr(*, "class")= chr "bigSNP"</span>

<span class="co"># Load the pvdiv phenotypes into the R session.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pvdiv_phenotypes</span><span class="op">)</span> 
<span class="co"># Make an example dataframe of one phenotype where the first column is PLANT_ID.</span>
<span class="co"># This "phenotype", 'GWAS_CT', is the number of times a plant successfully </span>
<span class="co"># clonally replicated to plant in the common gardens in 2018.</span>
<span class="va">one_phenotype</span> <span class="op">&lt;-</span> <span class="va">pvdiv_phenotypes</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">PLANT_ID</span>, <span class="va">GWAS_CT</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-genome-wide-association-gwas">Run genome-wide association (GWAS)<a class="anchor" aria-label="anchor" href="#run-genome-wide-association-gwas"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Save the output to a temporary directory for this demo.</span>
<span class="va">tempdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas</a></span><span class="op">(</span>snp <span class="op">=</span> <span class="va">snp_example</span>, df <span class="op">=</span> <span class="va">one_phenotype</span>, 
                    type <span class="op">=</span> <span class="st">"linear"</span>, outputdir <span class="op">=</span> <span class="va">tempdir</span>, 
                    savegwas <span class="op">=</span> <span class="cn">FALSE</span>, saveplots <span class="op">=</span> <span class="cn">TRUE</span>,
                    saveannos <span class="op">=</span> <span class="cn">FALSE</span>, ncores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; 'lambdagc' is TRUE, so lambda_GC will be used to find the best population structure correction using the covariance matrix.</span>
<span class="co">#&gt; 'savegwas' is FALSE, so the gwas results will not be saved to disk.</span>
<span class="co">#&gt; Covariance matrix (covar) was not supplied - this will be generated using pvdiv_autoSVD().</span>
<span class="co">#&gt; 'saveoutput' is FALSE, so the svd will not be saved to the working directory.</span>
<span class="co">#&gt; Now starting GWAS pipeline for GWAS_CT.</span>
<span class="co">#&gt; Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="co">#&gt; saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 0 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 1 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 2 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 3 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 4 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 5 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 6 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 7 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 8 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 9 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 10 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 11 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 12 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 13 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 14 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 15 PCs.</span>
<span class="co">#&gt; Finished phenotype 1: GWAS_CT</span>
<span class="co">#&gt; Now running GWAS with the best population structure correction.</span>
<span class="co">#&gt; [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="co">#&gt; Now generating and saving Manhattan and QQ plots.</span></code></pre></div>
<p>This command will save a Manhattan and QQ-plot for the ~1800 SNPs from the example file to a temporary directory, which will have a randomly generated name like: C:3b2.</p>
<p>The example Manhattan plot should look like this:</p>
<div class="figure">
<img src="../reference/figures/Ex_Manhattan.png" style="width:100.0%" alt=""><p class="caption">Manhattan</p>
</div>
<p>And the example QQ plot should look like this.</p>
<div class="figure">
<img src="../reference/figures/Ex_QQ.png" style="width:50.0%" alt=""><p class="caption">QQ-plot</p>
</div>
<p><code>pvdiv_standard_gwas</code> is a wrapper function for the standard GWAS done in the Juenger lab. You must specify three things to use this function: 1) a snp file in bigSNP format, 2) a phenotype file where the first column is PLANT_ID (as in <code><a href="../reference/pvdiv_metadata.html">switchgrassGWAS::pvdiv_metadata</a></code>), and 3) whether the GWAS should be a linear or logistic regression.</p>
<div class="section level3">
<h3 id="additional-features">Additional features:<a class="anchor" aria-label="anchor" href="#additional-features"></a>
</h3>
<pre><code>* Specify `savegwas = TRUE` if you want the GWAS outputs to be saved to the output directory as .rds files.
* Specify `saveplots = TRUE` if you want Manhattan and QQ-plots to be generated and saved to the output directory. (NB: This is the default.)
* Specify `saveannos = TRUE` and specify a txdb object loaded into your environment with AnnotationDbi::loadDb, if you want annotation tables for top SNPs to be generated and saved to the output directory.
* Specify an integer value for `minphe` if you want to specify a minimum number of phenotyped individuals to conduct a GWAS on. </code></pre>
</div>
<div class="section level3">
<h3 id="troubleshooting">Troubleshooting:<a class="anchor" aria-label="anchor" href="#troubleshooting"></a>
</h3>
<p>If you get the error “Error in snp_autoSVD(G = G, infos.chr = CHRN$CHRN, infos.pos = POS, ncores = ncores, : object ‘LD.wiki34’ not found”</p>
<p>Likely you have not called <code><a href="https://privefl.github.io/bigsnpr/" class="external-link">library(bigsnpr)</a></code> in your current R session. You may need to run <code>install.packages("bigsnpr")</code> and then <code><a href="https://privefl.github.io/bigsnpr/" class="external-link">library(bigsnpr)</a></code>, then try running the <code><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas()</a></code> command again.</p>
<p>If you cannot find the temporary directory you created (tempdir): Try setting the output directory (outputdir) to your current working directory using outputdir = “.” within <code><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas()</a></code>, or tempdir &lt;- “.” &amp; keeping the <code><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas()</a></code> function as-is, then try running the <code><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas()</a></code> command again.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
