<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ switchgrassGWAS</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-datasets" class="hasAnchor"><a href="#section-datasets" class="anchor"></a>Datasets</h2>
          <p class="section-desc"><p>Datasets associated with the pvdiv panel</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="pvdiv_metadata.html">pvdiv_metadata</a></code> </p>
        </td>
        <td><p>Metadata table for the Panicum virgatum diversity panel</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_phenotypes.html">pvdiv_phenotypes</a></code> </p>
        </td>
        <td><p>Phenotypes available for the Panicum virgatum diversity panel.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="anno_info.html">anno_info</a></code> </p>
        </td>
        <td><p>Annotation information from Pvirgatum_516_v5.1.annotation_info</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gff_gene.html">gff_gene</a></code> </p>
        </td>
        <td><p>GFF3 Gene information</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-genome-wide-association" class="hasAnchor"><a href="#section-genome-wide-association" class="anchor"></a>Genome-wide association</h2>
          <p class="section-desc"><p>Functions for genome-wide association on the pvdiv panel</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="pvdiv_standard_gwas.html">pvdiv_standard_gwas()</a></code> </p>
        </td>
        <td><p>Juenger lab standard GWAS function.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_gwas.html">pvdiv_gwas()</a></code> </p>
        </td>
        <td><p>Wrapper for bigsnpr for GWAS on Panicum virgatum.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_qqplot.html">pvdiv_qqplot()</a></code> </p>
        </td>
        <td><p>Create a quantile-quantile plot with ggplot2.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_table_topsnps.html">pvdiv_table_topsnps()</a></code> </p>
        </td>
        <td><p>Create a Table of Annotated Top SNPs for Panicum virgatum</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_results_in_folder.html">pvdiv_results_in_folder()</a></code> </p>
        </td>
        <td><p>Identify phenotype names from bigsnpr results in a folder.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_kinship.html">pvdiv_kinship()</a></code> </p>
        </td>
        <td><p>Find a kinship matrix using the van Raden method.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_lambda_GC.html">pvdiv_lambda_GC()</a></code> </p>
        </td>
        <td><p>Return lambda_GC for different numbers of PCs for GWAS on Panicum virgatum.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-multivariate-adaptive-shrinkage" class="hasAnchor"><a href="#section-multivariate-adaptive-shrinkage" class="anchor"></a>Multivariate adaptive shrinkage</h2>
          <p class="section-desc"><p>Functions for multi-condition SNP effect estimates using mash</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="pvdiv_results_in_folder.html">pvdiv_results_in_folder()</a></code> </p>
        </td>
        <td><p>Identify phenotype names from bigsnpr results in a folder.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvdiv_bigsnp2mashr.html">pvdiv_bigsnp2mashr()</a></code> </p>
        </td>
        <td><p>Convert bigsnpr output to mashr input dataframes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_standard_run.html">mash_standard_run()</a></code> </p>
        </td>
        <td><p>A standard run of mashr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_plot_Ulist.html">mash_plot_Ulist()</a></code> </p>
        </td>
        <td><p>ggplot of specific covariance matrix patterns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_plot_covar.html">mash_plot_covar()</a></code> </p>
        </td>
        <td><p>ggplot of covariance matrix masses</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_plot_effects.html">mash_plot_effects()</a></code> </p>
        </td>
        <td><p>ggplot of single mash effect</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_plot_manhattan_by_condition.html">mash_plot_manhattan_by_condition()</a></code> </p>
        </td>
        <td><p>Manhattan plot in ggplot colored by significant conditions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_plot_pairwise_sharing.html">mash_plot_pairwise_sharing()</a></code> </p>
        </td>
        <td><p>Create a ggplot of pairwise sharing of mash effects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mash_plot_sig_by_condition.html">mash_plot_sig_by_condition()</a></code> </p>
        </td>
        <td><p>Significant SNPs per number of conditions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_GxE.html">get_GxE()</a></code> </p>
        </td>
        <td><p>Get data frames of types of GxE from a mash object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_U_by_mass.html">get_U_by_mass()</a></code> </p>
        </td>
        <td><p>Get the positions of objects in a mash object Ulist that are above
some mass threshold.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_PC_df.html">get_best_PC_df()</a></code> </p>
        </td>
        <td><p>Return best number of PCs in terms of lambda_GC for Panicum virgatum.
Return best number of PCs in terms of lambda_GC for the CDBN.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_colnames.html">get_colnames()</a></code> </p>
        </td>
        <td><p>Get column names from a mash object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_date_filename.html">get_date_filename()</a></code> </p>
        </td>
        <td><p>Get current date-time in a filename-appropriate format.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_estimated_pi.html">get_estimated_pi()</a></code> </p>
        </td>
        <td><p>Return the estimated mixture proportions. Use get_estimated_pi to
extract the estimates of the mixture proportions for different types of
covariance matrix. This tells you which covariance matrices have most of
the mass.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_lambdagc.html">get_lambdagc()</a></code> </p>
        </td>
        <td><p>Find lambda_GC value for non-NA p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_log10bf.html">get_log10bf()</a></code> </p>
        </td>
        <td><p>Return the Bayes Factor for each effect</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_marker_df.html">get_marker_df()</a></code> </p>
        </td>
        <td><p>Get mash marker_df</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_mash_metadata.html">get_mash_metadata()</a></code> </p>
        </td>
        <td><p>Get switchgrass metadata for SNPs in a mash object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_n_significant_conditions.html">get_n_significant_conditions()</a></code> </p>
        </td>
        <td><p>Count number of conditions each effect is significant in</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ncond.html">get_ncond()</a></code> </p>
        </td>
        <td><p>Get number of conditions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_pairwise_sharing.html">get_pairwise_sharing()</a></code> </p>
        </td>
        <td><p>Compute the proportion of (significant) signals shared by magnitude in each pair of conditions, based on the poterior mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_significant_results.html">get_significant_results()</a></code> <code><a href="get_significant_results.html">get_significant_results()</a></code> </p>
        </td>
        <td><p>From a mash result, get effects that are significant in at least
one condition.</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


