<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using file-backed matrices - FBM • switchgrassGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using file-backed matrices - FBM">
<meta property="og:description" content="switchgrassGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/fbm.html">Using file-backed matrices - FBM</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/inputs.html">SNP file inputs and QC</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="fbm_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using file-backed matrices - FBM</h1>
                        <h4 data-toc-skip class="author">Alice MacQueen</h4>
            
            <h4 data-toc-skip class="date">2022-01-13</h4>
      
      
      <div class="hidden name"><code>fbm.Rmd</code></div>

    </div>

    
    
<p>When you run GWAS using this package, you now have two options for saving the GWAS results - as RDS files and as file-backed matrix (FBM) files.</p>
<p>You can specify this option in <code><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas()</a></code> with the option <code>savetype</code>. <code>savetype</code> can be “rds”, “fbm”, or “both”.</p>
<div class="section level3">
<h3 id="why-use-rds-files">Why use RDS files?<a class="anchor" aria-label="anchor" href="#why-use-rds-files"></a>
</h3>
<p>If you are running few GWAS and want to save each GWAS output separately, you can use rds files. These files completely describe the GWAS results without referring back to the original SNP file, using the columns CHR (chromosome) &amp; POS (physical position). They contain the effect estimate <code>estim</code> for the alternate allele (allele 2 in the SNP file), the standard error around this effect <code>std_err</code>, and the -log10pvalue and FDR adjusted p-values <code>log10p</code> and <code>FDR_adj</code> from the genome-wide association.</p>
<div class="section level4">
<h4 id="pros-of-rds-files">Pros of RDS files<a class="anchor" aria-label="anchor" href="#pros-of-rds-files"></a>
</h4>
<p>These files are standalone &amp; contain all the information you need to analyze a GWAS result.</p>
<p>These rds files can be used with the function pvdiv_bigsnp2mashr to run mash, as well as with pvdiv_table_topsnps (by specifying type = “bigsnp”), pvdiv_qqplot, and pvdiv_manhattan.</p>
</div>
<div class="section level4">
<h4 id="cons-of-rds-files">Cons of RDS files<a class="anchor" aria-label="anchor" href="#cons-of-rds-files"></a>
</h4>
<p>If you are running many GWAS, these files can become numerous, take up substantial disk space, and become hard to keep track of. Individual files make it more difficult to compare results across multiple GWAS.</p>
</div>
</div>
<div class="section level3">
<h3 id="why-use-fbm-files">Why use FBM files?<a class="anchor" aria-label="anchor" href="#why-use-fbm-files"></a>
</h3>
<p>If you are running many GWAS, you may want a compressed format, a reduced number of files, and/or additional ways to compare GWAS outputs. To do this, you can use FBM files, which save all GWAS results in a total of 4 files. These files all start with the prefix ‘gwas_effects_’. Two are .csv files; these contain the associated metadata for the GWAS and the column names for the fbm. Two are an .rds &amp; .bk format &amp; contain the actual file-backed matrix results; these can be very large.</p>
<p>For each phenotype used for GWAS, the gwas_effects files contain the effect estimate <code>estim</code> for the alternate allele (allele 2 in the SNP file), the standard error around this effect <code>std_err</code>, and the -log10pvalue <code>log10p</code> from the genome-wide association.</p>
<div class="section level4">
<h4 id="pros-of-fbm-files">Pros of FBM files<a class="anchor" aria-label="anchor" href="#pros-of-fbm-files"></a>
</h4>
<p>FBM files are memory and computation-efficient tools to analyze big matrices that are stored on disc rather than in your computer’s memory (<a href="https://privefl.github.io/R-presentation/bigstatsr.html#2" class="external-link">see here for more details</a> ). Briefly, they have faster I/O, won’t overwhelm your computer’s RAM, and function runtime speeds up ~10-200x.</p>
<p>These fbm files can be used with the functions pvdiv_fbm_upset_df() and pvdiv_fbm_qq(). They can also be used in the snpdiver R package function dive_effects2mash() to run mash.</p>
</div>
<div class="section level4">
<h4 id="cons-of-fbm-files">Cons of FBM files<a class="anchor" aria-label="anchor" href="#cons-of-fbm-files"></a>
</h4>
<p>These files need to be used along with the original snp file used to run the GWAS - thus, you’ll need to keep track of six files in total, the .rds and .bk file for the SNP file and the four files for the gwas results.</p>
<p>These files do not work with the function pvdiv_bigsnp2mashr to run mash.</p>
<p>If you are interested in only a subset of GWAS results saved in a fbm, you will have to create a vector of row numbers using the associated metadata to refer to that data subset.</p>
</div>
</div>
<div class="section level2">
<h2 id="using-fbm-gwas-results">Using FBM gwas results<a class="anchor" aria-label="anchor" href="#using-fbm-gwas-results"></a>
</h2>
<p>The remainder of this package demonstrates some utilities for filebacked matrix files within the bigsnpr and switchgrassGWAS R packages.</p>
<div class="section level3">
<h3 id="get-genotype-file">Get genotype file<a class="anchor" aria-label="anchor" href="#get-genotype-file"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the example bedfile from the package switchgrassGWAS</span>
<span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.bed"</span>, package <span class="op">=</span> <span class="st">"switchgrassGWAS"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="set-up-snp-and-phenotype-data-frames-">Set up SNP and phenotype data frames.<a class="anchor" aria-label="anchor" href="#set-up-snp-and-phenotype-data-frames-"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load packages bigsnpr and switchgrassGWAS</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alice-macqueen.github.io/switchgrassGWAS" class="external-link">switchgrassGWAS</a></span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span>
<span class="co">#&gt;   method from   </span>
<span class="co">#&gt;   +.gg   ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: bigstatsr</span>

<span class="co"># Read from bed/bim/fam to create the new files that bigsnpr uses.</span>
<span class="co"># Let's put them in an temporary directory for this demo.</span>
<span class="va">tmpfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="va">tmpfile</span><span class="op">)</span>
<span class="co">#&gt; [1] "/tmp/Rtmp9mfsVL/file2f6749d8f793.rds"</span>

<span class="co"># Attach the "bigSNP" object to the R session.</span>
<span class="va">snp_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpfile</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># What does the bigSNP object look like?</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">snp_example</span>, max.level <span class="op">=</span> <span class="fl">2</span>, strict.width <span class="op">=</span> <span class="st">"cut"</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 fields</span>
<span class="co">#&gt;   ..and 26 methods, of which 12 are  possibly relevant:</span>
<span class="co">#&gt;   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span>
<span class="co">#&gt;   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,</span>
<span class="co">#&gt;   ..  save, show#envRefClass, show#FBM</span>
<span class="co">#&gt;  $ fam      :'data.frame':   630 obs. of  6 variables:</span>
<span class="co">#&gt;   ..$ family.ID  : chr [1:630] "Pvirgatum" "Pvirgatum" "Pvirgatum" "Pvirgatum"..</span>
<span class="co">#&gt;   ..$ sample.ID  : chr [1:630] "J181.A" "J250.C" "J251.C" "J352.A" ...</span>
<span class="co">#&gt;   ..$ paternal.ID: int [1:630] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ maternal.ID: int [1:630] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ sex        : int [1:630] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ affection  : int [1:630] -9 -9 -9 -9 -9 -9 -9 -9 -9 -9 ...</span>
<span class="co">#&gt;  $ map      :'data.frame':   3600 obs. of  6 variables:</span>
<span class="co">#&gt;   ..$ chromosome  : chr [1:3600] "Chr01K" "Chr01K" "Chr01K" "Chr01K" ...</span>
<span class="co">#&gt;   ..$ marker.ID   : chr [1:3600] "Chr01K_105542" "Chr01K_798650" "Chr01K_1147"..</span>
<span class="co">#&gt;   ..$ genetic.dist: int [1:3600] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ physical.pos: int [1:3600] 105542 798650 1147074 1719551 1915851 1965986..</span>
<span class="co">#&gt;   ..$ allele1     : chr [1:3600] "A" "C" "T" "T" ...</span>
<span class="co">#&gt;   ..$ allele2     : chr [1:3600] "G" "G" "C" "G" ...</span>
<span class="co">#&gt;  - attr(*, "class")= chr "bigSNP"</span>

<span class="co"># Load the pvdiv phenotypes into the R session.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pvdiv_phenotypes</span><span class="op">)</span> 
<span class="co"># Make an example dataframe of one phenotype where the first column is PLANT_ID.</span>
<span class="co"># This "phenotype", 'GWAS_CT', is the number of times a plant successfully </span>
<span class="co"># clonally replicated to plant in the common gardens in 2018.</span>
<span class="va">four_phenotype</span> <span class="op">&lt;-</span> <span class="va">pvdiv_phenotypes</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">PLANT_ID</span>, <span class="va">GWAS_CT</span>, <span class="fu">matches</span><span class="op">(</span><span class="st">"BIOMASS"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-genome-wide-association-gwas">Run genome-wide association (GWAS)<a class="anchor" aria-label="anchor" href="#run-genome-wide-association-gwas"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Save the output to a temporary directory for this demo.</span>
<span class="va">tempdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="../reference/pvdiv_standard_gwas.html">pvdiv_standard_gwas</a></span><span class="op">(</span>snp <span class="op">=</span> <span class="va">snp_example</span>, df <span class="op">=</span> <span class="va">four_phenotype</span>, 
                    type <span class="op">=</span> <span class="st">"linear"</span>, outputdir <span class="op">=</span> <span class="va">tempdir</span>, 
                    savegwas <span class="op">=</span> <span class="cn">TRUE</span>, savetype <span class="op">=</span> <span class="st">"fbm"</span>, 
                    saveplots <span class="op">=</span> <span class="cn">FALSE</span>,
                    saveannos <span class="op">=</span> <span class="cn">FALSE</span>, ncores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; 'lambdagc' is TRUE, so lambda_GC will be used to find the best population structure correction using the covariance matrix.</span>
<span class="co">#&gt; 'savetype' is 'fbm', so the gwas results will be saved to disk as a combined filebacked matrix (fbm) file.</span>
<span class="co">#&gt; Covariance matrix (covar) was not supplied - this will be generated using pvdiv_autoSVD().</span>
<span class="co">#&gt; 'saveoutput' is FALSE, so the svd will not be saved to the working directory.</span>
<span class="co">#&gt; Now starting GWAS pipeline for GWAS_CT.</span>
<span class="co">#&gt; Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="co">#&gt; saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 0 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 1 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 2 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 3 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 4 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 5 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 6 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 7 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 8 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 9 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 10 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 11 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 12 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 13 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 14 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for GWAS_CT using 15 PCs.</span>
<span class="co">#&gt; Finished phenotype 1: GWAS_CT</span>
<span class="co">#&gt; Now running GWAS with the best population structure correction.</span>
<span class="co">#&gt; [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="co">#&gt; Now starting GWAS pipeline for CLMB_BIOMASS.</span>
<span class="co">#&gt; Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="co">#&gt; saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 0 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 1 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 2 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 3 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 4 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 5 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 6 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 7 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 8 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 9 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 10 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 11 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 12 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 13 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 14 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for CLMB_BIOMASS using 15 PCs.</span>
<span class="co">#&gt; Finished phenotype 1: CLMB_BIOMASS</span>
<span class="co">#&gt; Now running GWAS with the best population structure correction.</span>
<span class="co">#&gt; [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="co">#&gt; Now starting GWAS pipeline for KBSM_BIOMASS.</span>
<span class="co">#&gt; Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="co">#&gt; saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 0 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 1 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 2 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 3 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 4 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 5 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 6 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 7 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 8 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 9 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 10 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 11 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 12 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 13 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 14 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for KBSM_BIOMASS using 15 PCs.</span>
<span class="co">#&gt; Finished phenotype 1: KBSM_BIOMASS</span>
<span class="co">#&gt; Now running GWAS with the best population structure correction.</span>
<span class="co">#&gt; [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span>
<span class="co">#&gt; Now starting GWAS pipeline for PKLE_BIOMASS.</span>
<span class="co">#&gt; Now determining lambda_GC for GWAS models with 16 sets of PCs. This will take some time.</span>
<span class="co">#&gt; saveoutput is FALSE, so lambda_GC values won't be saved to a csv.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 0 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 1 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 2 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 3 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 4 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 5 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 6 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 7 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 8 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 9 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 10 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 11 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 12 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 13 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 14 PCs.</span>
<span class="co">#&gt; Finished Lambda_GC calculation for PKLE_BIOMASS using 15 PCs.</span>
<span class="co">#&gt; Finished phenotype 1: PKLE_BIOMASS</span>
<span class="co">#&gt; Now running GWAS with the best population structure correction.</span>
<span class="co">#&gt; [1] "saveoutput is FALSE so GWAS object will not be saved to disk."</span></code></pre></div>
<p>This command will save a set of gwas results to a temporary directory, which will have a randomly generated name like: /tmp/Rtmp9mfsVL.</p>
<p>Let’s load these files into R.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_attach.html" class="external-link">big_attach</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tempdir</span>, <span class="st">"gwas_effects_0M_SNPs.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gwas_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tempdir</span>, <span class="st">"gwas_effects_0M_SNPs_associated_metadata.csv"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gwas_colnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tempdir</span>, <span class="st">"gwas_effects_0M_SNPs_column_names.csv"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Let’s look at the metadata associated with the GWAS. This includes information about the kind of GWAS that was conducted. In order, the columns indicate, what the phenotype name was, what type of model was run, how many SNPs (in M) were used, how many PCs were used for population structure correction, how many phenotyped individuals were there, how many phenotypic levels were there, what was the genomic inflation factor (lambda_GC), and given the number of SNPs, what was the bonferroni threshold for the phenotype.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">kable</span><span class="op">(</span><span class="va">gwas_metadata</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">phe</th>
<th align="left">type</th>
<th align="left">nsnp</th>
<th align="right">npcs</th>
<th align="right">nphe</th>
<th align="right">nlev</th>
<th align="right">lambda_GC</th>
<th align="right">bonferroni</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">GWAS_CT</td>
<td align="left">linear</td>
<td align="left">0_M</td>
<td align="right">15</td>
<td align="right">629</td>
<td align="right">16</td>
<td align="right">1.12682</td>
<td align="right">4.857333</td>
</tr>
<tr class="even">
<td align="left">CLMB_BIOMASS</td>
<td align="left">linear</td>
<td align="left">0_M</td>
<td align="right">5</td>
<td align="right">495</td>
<td align="right">490</td>
<td align="right">1.02958</td>
<td align="right">4.857333</td>
</tr>
<tr class="odd">
<td align="left">KBSM_BIOMASS</td>
<td align="left">linear</td>
<td align="left">0_M</td>
<td align="right">15</td>
<td align="right">512</td>
<td align="right">512</td>
<td align="right">1.07005</td>
<td align="right">4.857333</td>
</tr>
<tr class="even">
<td align="left">PKLE_BIOMASS</td>
<td align="left">linear</td>
<td align="left">0_M</td>
<td align="right">10</td>
<td align="right">612</td>
<td align="right">606</td>
<td align="right">1.03844</td>
<td align="right">4.857333</td>
</tr>
</tbody>
</table>
<p>Let’s look at the column names of the GWAS result. These are in the same order as the associated metadata, and each phenotype has three columns: Effect (estimated effect size of allele2), SE (standard error about this effect size), and log10p (-log10p of the association between SNP &amp; phenotype).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">kable</span><span class="op">(</span><span class="va">gwas_colnames</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">colnames_fbm</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">GWAS_CT_Effect</td>
</tr>
<tr class="even">
<td align="left">GWAS_CT_SE</td>
</tr>
<tr class="odd">
<td align="left">GWAS_CT_log10p</td>
</tr>
<tr class="even">
<td align="left">CLMB_BIOMASS_Effect</td>
</tr>
<tr class="odd">
<td align="left">CLMB_BIOMASS_SE</td>
</tr>
<tr class="even">
<td align="left">CLMB_BIOMASS_log10p</td>
</tr>
<tr class="odd">
<td align="left">KBSM_BIOMASS_Effect</td>
</tr>
<tr class="even">
<td align="left">KBSM_BIOMASS_SE</td>
</tr>
<tr class="odd">
<td align="left">KBSM_BIOMASS_log10p</td>
</tr>
<tr class="even">
<td align="left">PKLE_BIOMASS_Effect</td>
</tr>
<tr class="odd">
<td align="left">PKLE_BIOMASS_SE</td>
</tr>
<tr class="even">
<td align="left">PKLE_BIOMASS_log10p</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="two-gwas-comparisons">Two-GWAS Comparisons<a class="anchor" aria-label="anchor" href="#two-gwas-comparisons"></a>
</h3>
<p>Let’s analyze overlaps in significance in the four GWAS. Is there a relationship between the -log10pvalues in the three BIOMASS GWAS? If so, this could indicate that SNPs have a similar effect on BIOMASS across the three different environments.</p>
<p>Use the row numbers from the associated metadata to compare -log10p values for row 2 (CLMB_BIOMASS) with rows 3 &amp; 4 (biomass at KBSM &amp; PKLE). Set the threshold above which to plot the p-value at zero (default is 5), because there are not many SNPs in this toy example.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pvdiv_fbm_qq.html">pvdiv_fbm_qq</a></span><span class="op">(</span>effects <span class="op">=</span> <span class="va">gwas</span>, metadata <span class="op">=</span> <span class="va">gwas_metadata</span>, e_row <span class="op">=</span> <span class="fl">2</span>, o_row <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">4</span>, 
             outputdir <span class="op">=</span> <span class="va">tempdir</span>, thr <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span>
<span class="co">#&gt; `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre></div>
<p>This function will again save plots to the tempdir. These plots should look like this:</p>
<div class="figure">
<img src="../reference/figures/GWAS_significance_thr_log10p_0_CLMB_BIOMASS_vs_KBSM_BIOMASS.png" style="width:50.0%" alt=""><p class="caption">CLMB_vs_KBSM</p>
</div>
<p>and this:</p>
<div class="figure">
<img src="../reference/figures/GWAS_significance_thr_log10p_0_CLMB_BIOMASS_vs_PKLE_BIOMASS.png" style="width:50.0%" alt=""><p class="caption">CLMB_vs_PKLE</p>
</div>
<p>If these GWAS results were perfectly related, all the hexagons would fall on the red dashed 1:1 line. The blue line indicates the actual relationship between the p-values in the two GWAS. We can see that there is a modest positive relationship between biomass in these environments, but it’s by no means strong. This indicates the presence of genotype-by-environment interactions between these environments.</p>
</div>
<div class="section level3">
<h3 id="upset-plots">Upset plots<a class="anchor" aria-label="anchor" href="#upset-plots"></a>
</h3>
<p>Now, let’s compare results for more than one GWAS. The way I recommend doing this is with an <a href="https://upset.app/" class="external-link">Upset plot</a>. Upset plots are well suited to the quantitative analysis of data when there are more than three sets. They shine when you want to look at all combinations of how sets interact - for example, when you’re interested in sets of SNPs significant in different combinations of GWAS.</p>
<p>Here’s some great documentation on <a href="https://krassowski.github.io/complex-upset/articles/Examples_R.html" class="external-link">how to code an Upset plot in R</a>.</p>
<p>Let’s make a dataframe to go into an Upset plot and visualize it in R.</p>
<p>There are three required arguments:</p>
<ul>
<li>the first argument is expected to be the gwas results fbm</li>
<li>the second argument is expected to be the snp fbm used to generate the gwas results</li>
<li>the third argument is expected to be the metadata associated with the GWAS results.</li>
</ul>
<p>We can also optionally specify a -log10p threshold with ‘thr’. By default, that threshold is 7. A good target is to set this above the noise floor for all of the GWAS you are comparing. For our toy example, we set this artificially low, to 2.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">upset_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pvdiv_fbm_upset_df.html">pvdiv_fbm_upset_df</a></span><span class="op">(</span>effects <span class="op">=</span> <span class="va">gwas</span>, snp <span class="op">=</span> <span class="va">snp_example</span>, 
                               metadata <span class="op">=</span> <span class="va">gwas_metadata</span>, thr <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">upset_df</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 343 × 6</span></span>
<span class="co">#&gt;    CHR         POS GWAS_CT CLMB_BIOMASS KBSM_BIOMASS PKLE_BIOMASS</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Chr01K  1<span style="text-decoration: underline;">147</span>074       0            0            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Chr01K  4<span style="text-decoration: underline;">761</span>900       0            0            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Chr01K  5<span style="text-decoration: underline;">505</span>667       0            0            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Chr01K  7<span style="text-decoration: underline;">103</span>018       0            0            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Chr01K  7<span style="text-decoration: underline;">475</span>083       0            0            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Chr01K  8<span style="text-decoration: underline;">046</span>966       0            0            1            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Chr01K 10<span style="text-decoration: underline;">626</span>578       0            0            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Chr01K 16<span style="text-decoration: underline;">537</span>641       0            1            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Chr01K 16<span style="text-decoration: underline;">575</span>349       0            1            0            1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Chr01K 18<span style="text-decoration: underline;">316</span>176       0            1            0            0</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 333 more rows</span></span></code></pre></div>
<p>This dataframe has CHR &amp; POS from the SNP file, and then an additional column for each phenotype. If that SNP is not significant at the -log10p value threshold for that phenotype (here, <code>thr = 2</code>), then that row &amp; column combination will have a ‘0’. If that SNP is significant for the phenotype, that value will be ‘1’.</p>
<p>Now, let’s make an upset plot using the R package ComplexUpset.</p>
<p>There are two required arguments:</p>
<ul>
<li>the first argument is expected to be a dataframe with both group indicator variables and covariates,</li>
<li>the second argument specifies a list with names of column which indicate the group membership.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/krassowski/complex-upset" class="external-link">ComplexUpset</a></span><span class="op">)</span>
<span class="va">upset_plot</span> <span class="op">&lt;-</span> <span class="fu">upset</span><span class="op">(</span><span class="va">upset_df</span>, intersect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">upset_df</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span>, 
                    name <span class="op">=</span> <span class="st">'phenotype'</span>, min_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">upset_plot</span>                    </code></pre></div>
<p><img src="../reference/figures/Ex_Upset.png" style="width:50.0%" alt="upset_ex"> In this Upset plot, we can see that there are many more SNPs significant in one or more biomass phenotype than there are significant for the phenotype ‘GWAS_CT’ and any biomass phenotype. Though, overall, most SNPs are significant in only one of the four conditions.</p>
<p>Sometimes this could be because different SNPs in the same gene or genomic region are coming up as highly associated with the different phenotypes. To test this, we can summarize our upset_df slightly based on position, and make a new Upset plot.</p>
<p>Median LD in switchgrass is between 20 and 25kb. Normally this is the largest genomic region I would make. However, for this toy example, we’ll make larger genomic regions of 1Mb.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>

<span class="va">upset_df_1Mb</span> <span class="op">&lt;-</span> <span class="va">upset_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>POS_binned <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">POS</span><span class="op">/</span><span class="fl">1000000</span><span class="op">)</span><span class="op">*</span><span class="fl">1000000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">POS</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">CHR</span>, <span class="va">POS_binned</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">CHR</span>, <span class="va">POS_binned</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; `summarise()` has grouped output by 'CHR'. You can override using the `.groups` argument.</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">upset_1Mb</span> <span class="op">&lt;-</span> <span class="fu">upset</span><span class="op">(</span><span class="va">upset_df_1Mb</span>, intersect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-toTable.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">upset_df_1Mb</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span>, 
                    name <span class="op">=</span> <span class="st">'phenotype'</span>, min_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Significant at -log10p &gt; 2 in 1Mb intervals"</span><span class="op">)</span>
<span class="va">upset_1Mb</span>  </code></pre></div>
<div class="figure">
<img src="../reference/figures/Ex_Upset_1Mb.png" style="width:50.0%" alt=""><p class="caption">upset_ex</p>
</div>
<p>This binning by genomic interval does result in more genomic regions with significant effects on multiple phenotypes. It doesn’t change the ordering of the kinds of intersections very much, with the exception that the three-way intersection between the three biomass phenotypes is now the least common intersection (with 4 1Mb regions containing significant associations for all three phenotypes).</p>
<p>What if we want to learn more about these 4 1Mb regions? We can go back to our upset dataframes to extract these regions.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sel_reg_biomass</span> <span class="op">&lt;-</span> <span class="va">upset_df_1Mb</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="va">CLMB_BIOMASS</span><span class="op">:</span><span class="va">PKLE_BIOMASS</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Sel_region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">CHR</span>, <span class="va">POS_binned</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span>

<span class="va">sel_snp_biomass</span> <span class="op">&lt;-</span> <span class="va">upset_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>POS_binned <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">POS</span><span class="op">/</span><span class="fl">1000000</span><span class="op">)</span><span class="op">*</span><span class="fl">1000000</span>,
         Sel_region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">CHR</span>, <span class="va">POS_binned</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Sel_region</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">sel_reg_biomass</span><span class="op">$</span><span class="va">Sel_region</span><span class="op">)</span>

<span class="fu">kable</span><span class="op">(</span><span class="va">sel_snp_biomass</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CHR</th>
<th align="right">POS</th>
<th align="right">GWAS_CT</th>
<th align="right">CLMB_BIOMASS</th>
<th align="right">KBSM_BIOMASS</th>
<th align="right">PKLE_BIOMASS</th>
<th align="right">POS_binned</th>
<th align="left">Sel_region</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Chr03K</td>
<td align="right">2197885</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2.0e+06</td>
<td align="left">Chr03K_2e+06</td>
</tr>
<tr class="even">
<td align="left">Chr05K</td>
<td align="right">12177803</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1.2e+07</td>
<td align="left">Chr05K_1.2e+07</td>
</tr>
<tr class="odd">
<td align="left">Chr09K</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
</tr>
<tr class="even">
<td align="left">Chr09N</td>
<td align="right">11325811</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
</tr>
<tr class="odd">
<td align="left">Chr09N</td>
<td align="right">11967533</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
</tr>
</tbody>
</table>
<p>If we were interested enough in these regions, we could also use this table to find annotations using <code><a href="../reference/pvdiv_table_topsnps.html">pvdiv_table_topsnps()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sel_snp_annos</span> <span class="op">&lt;-</span> <span class="va">sel_snp_biomass</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>start <span class="op">=</span> <span class="va">POS</span> <span class="op">-</span> <span class="fl">10000</span>,
         end <span class="op">=</span> <span class="va">POS</span> <span class="op">+</span> <span class="fl">10000</span><span class="op">)</span>  <span class="co"># get the 20kb surrounding each position, POS, in thr_dfB</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi" class="external-link">AnnotationDbi</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"../../pvdiv-genome/Pvirgatum_516_v5.1.gene.txdb.sqlite"</span><span class="op">)</span>

<span class="va">sel_snp_annos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pvdiv_table_topsnps.html">pvdiv_table_topsnps</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">sel_snp_annos</span>, type <span class="op">=</span> <span class="st">"table"</span>, txdb <span class="op">=</span> <span class="va">txdb</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">kable</span><span class="op">(</span><span class="va">sel_snp_annos</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="1%">
<col width="1%">
<col width="0%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="3%">
<col width="3%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="9%">
<col width="4%">
<col width="3%">
<col width="12%">
<col width="2%">
<col width="2%">
<col width="11%">
<col width="1%">
<col width="1%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="left">CHR</th>
<th align="right">region_start</th>
<th align="right">region_end</th>
<th align="right">width</th>
<th align="left">region_strand</th>
<th align="left">Annotation</th>
<th align="right">QUERYID</th>
<th align="left">TXID</th>
<th align="left">Gene ID</th>
<th align="right">POS</th>
<th align="right">GWAS_CT</th>
<th align="right">CLMB_BIOMASS</th>
<th align="right">KBSM_BIOMASS</th>
<th align="right">PKLE_BIOMASS</th>
<th align="right">POS_binned</th>
<th align="left">Sel_region</th>
<th align="right">gene_start</th>
<th align="right">gene_end</th>
<th align="right">gene_width</th>
<th align="left">gene_strand</th>
<th align="right">pacId</th>
<th align="left">transcriptName</th>
<th align="left">peptideName</th>
<th align="left">Pfam</th>
<th align="left">Panther Categories</th>
<th align="left">KOG</th>
<th align="left">ec</th>
<th align="left">KO</th>
<th align="left">GO</th>
<th align="left">Arabidopsis thaliana homolog</th>
<th align="left">A. thaliana gene name</th>
<th align="left">A. thaliana gene annotation</th>
<th align="left">Rice homolog</th>
<th align="left">Rice gene name</th>
<th align="left">Rice gene annotation</th>
<th align="right">d2_start</th>
<th align="right">d2_end</th>
<th align="right">distance from gene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Chr03K</td>
<td align="right">2187885</td>
<td align="right">2207885</td>
<td align="right">20001</td>
<td align="left">*</td>
<td align="left">intergenic</td>
<td align="right">1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">2197885</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2.0e+06</td>
<td align="left">Chr03K_2e+06</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">Chr05K</td>
<td align="right">12167803</td>
<td align="right">12187803</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">2</td>
<td align="left">60175</td>
<td align="left">Pavir.5KG164407</td>
<td align="right">12177803</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1.2e+07</td>
<td align="left">Chr05K_1.2e+07</td>
<td align="right">12176848</td>
<td align="right">12177263</td>
<td align="right">416</td>
<td align="left">+</td>
<td align="right">41543573</td>
<td align="left">Pavir.5KG164407.1</td>
<td align="left">Pavir.5KG164407.1.p</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">955</td>
<td align="right">540</td>
<td align="right">540</td>
</tr>
<tr class="odd">
<td align="left">Chr05K</td>
<td align="right">12167803</td>
<td align="right">12187803</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">2</td>
<td align="left">60176</td>
<td align="left">Pavir.5KG164807</td>
<td align="right">12177803</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1.2e+07</td>
<td align="left">Chr05K_1.2e+07</td>
<td align="right">12185636</td>
<td align="right">12186204</td>
<td align="right">569</td>
<td align="left">+</td>
<td align="right">41549658</td>
<td align="left">Pavir.5KG164807.1</td>
<td align="left">Pavir.5KG164807.1.p</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">-7833</td>
<td align="right">-8401</td>
<td align="right">7833</td>
</tr>
<tr class="even">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">-</td>
<td align="left">intron</td>
<td align="right">3</td>
<td align="left">103241</td>
<td align="left">Pavir.8KG390800</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">41644684</td>
<td align="left">Pavir.8KG390800.1</td>
<td align="left">Pavir.8KG390800.1.p</td>
<td align="left">PF03810</td>
<td align="left">PTHR10997,PTHR10997:SF29</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0008536,GO:0006886" class="uri">GO:0008536,GO:0006886</a></td>
<td align="left">AT3G17340.1</td>
<td align="left">NA</td>
<td align="left">ARM repeat superfamily protein</td>
<td align="left">LOC_Os11g47320.1</td>
<td align="left">NA</td>
<td align="left">protein transporter, putative, expressed</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">3</td>
<td align="left">110376</td>
<td align="left">Pavir.9KG334200</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">45586592</td>
<td align="right">45589853</td>
<td align="right">3262</td>
<td align="left">+</td>
<td align="right">41604637</td>
<td align="left">Pavir.9KG334200.1</td>
<td align="left">Pavir.9KG334200.1.p</td>
<td align="left">PF00249</td>
<td align="left">PTHR24078,PTHR24078:SF195</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT5G47390.1</td>
<td align="left">NA</td>
<td align="left">myb-like transcription factor family protein</td>
<td align="left">LOC_Os10g41200.1</td>
<td align="left">NA</td>
<td align="left">MYB family transcription factor, putative, expressed</td>
<td align="right">968</td>
<td align="right">-2293</td>
<td align="right">968</td>
</tr>
<tr class="even">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">3</td>
<td align="left">116062</td>
<td align="left">Pavir.9KG334400</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">45576944</td>
<td align="right">45582938</td>
<td align="right">5995</td>
<td align="left">-</td>
<td align="right">41606480</td>
<td align="left">Pavir.9KG334400.4</td>
<td align="left">Pavir.9KG334400.4.p</td>
<td align="left">PF00582,PF04564,PF00069</td>
<td align="left">PTHR27003,PTHR27003:SF29</td>
<td align="left">KOG1187</td>
<td align="left">2.7.11.1</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0006950,GO:0016567,GO:0004842,GO:0006468,GO:0005524,GO:0004672" class="uri">GO:0006950,GO:0016567,GO:0004842,GO:0006468,GO:0005524,GO:0004672</a></td>
<td align="left">AT4G25160.1</td>
<td align="left">NA</td>
<td align="left">U-box domain-containing protein kinase family protein</td>
<td align="left">LOC_Os10g41220.1</td>
<td align="left">NA</td>
<td align="left">protein kinase family protein, putative, expressed</td>
<td align="right">10616</td>
<td align="right">4622</td>
<td align="right">4622</td>
</tr>
<tr class="odd">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">3</td>
<td align="left">116063</td>
<td align="left">Pavir.9KG334400</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">45576944</td>
<td align="right">45582938</td>
<td align="right">5995</td>
<td align="left">-</td>
<td align="right">41606480</td>
<td align="left">Pavir.9KG334400.4</td>
<td align="left">Pavir.9KG334400.4.p</td>
<td align="left">PF00582,PF04564,PF00069</td>
<td align="left">PTHR27003,PTHR27003:SF29</td>
<td align="left">KOG1187</td>
<td align="left">2.7.11.1</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0006950,GO:0016567,GO:0004842,GO:0006468,GO:0005524,GO:0004672" class="uri">GO:0006950,GO:0016567,GO:0004842,GO:0006468,GO:0005524,GO:0004672</a></td>
<td align="left">AT4G25160.1</td>
<td align="left">NA</td>
<td align="left">U-box domain-containing protein kinase family protein</td>
<td align="left">LOC_Os10g41220.1</td>
<td align="left">NA</td>
<td align="left">protein kinase family protein, putative, expressed</td>
<td align="right">10616</td>
<td align="right">4622</td>
<td align="right">4622</td>
</tr>
<tr class="even">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">3</td>
<td align="left">116064</td>
<td align="left">Pavir.9KG334100</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">45590557</td>
<td align="right">45594099</td>
<td align="right">3543</td>
<td align="left">-</td>
<td align="right">41607551</td>
<td align="left">Pavir.9KG334100.1</td>
<td align="left">Pavir.9KG334100.1.p</td>
<td align="left">PF07690</td>
<td align="left">PTHR23500,PTHR23500:SF179</td>
<td align="left">KOG0254</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0055085,GO:0016021" class="uri">GO:0055085,GO:0016021</a></td>
<td align="left">AT5G61520.1</td>
<td align="left">NA</td>
<td align="left">Major facilitator superfamily protein</td>
<td align="left">LOC_Os10g41190.1</td>
<td align="left">NA</td>
<td align="left">transporter family protein, putative, expressed</td>
<td align="right">-2997</td>
<td align="right">-6539</td>
<td align="right">2997</td>
</tr>
<tr class="odd">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">3</td>
<td align="left">116065</td>
<td align="left">Pavir.9KG334100</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">45590557</td>
<td align="right">45594099</td>
<td align="right">3543</td>
<td align="left">-</td>
<td align="right">41607551</td>
<td align="left">Pavir.9KG334100.1</td>
<td align="left">Pavir.9KG334100.1.p</td>
<td align="left">PF07690</td>
<td align="left">PTHR23500,PTHR23500:SF179</td>
<td align="left">KOG0254</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0055085,GO:0016021" class="uri">GO:0055085,GO:0016021</a></td>
<td align="left">AT5G61520.1</td>
<td align="left">NA</td>
<td align="left">Major facilitator superfamily protein</td>
<td align="left">LOC_Os10g41190.1</td>
<td align="left">NA</td>
<td align="left">transporter family protein, putative, expressed</td>
<td align="right">-2997</td>
<td align="right">-6539</td>
<td align="right">2997</td>
</tr>
<tr class="even">
<td align="left">Chr09K</td>
<td align="right">45577560</td>
<td align="right">45597560</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">3</td>
<td align="left">116066</td>
<td align="left">Pavir.9KG334100</td>
<td align="right">45587560</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4.5e+07</td>
<td align="left">Chr09K_4.5e+07</td>
<td align="right">45590557</td>
<td align="right">45594099</td>
<td align="right">3543</td>
<td align="left">-</td>
<td align="right">41607551</td>
<td align="left">Pavir.9KG334100.1</td>
<td align="left">Pavir.9KG334100.1.p</td>
<td align="left">PF07690</td>
<td align="left">PTHR23500,PTHR23500:SF179</td>
<td align="left">KOG0254</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0055085,GO:0016021" class="uri">GO:0055085,GO:0016021</a></td>
<td align="left">AT5G61520.1</td>
<td align="left">NA</td>
<td align="left">Major facilitator superfamily protein</td>
<td align="left">LOC_Os10g41190.1</td>
<td align="left">NA</td>
<td align="left">transporter family protein, putative, expressed</td>
<td align="right">-2997</td>
<td align="right">-6539</td>
<td align="right">2997</td>
</tr>
<tr class="odd">
<td align="left">Chr09N</td>
<td align="right">11315811</td>
<td align="right">11335811</td>
<td align="right">20001</td>
<td align="left">-</td>
<td align="left">intron</td>
<td align="right">4</td>
<td align="left">117033</td>
<td align="left">Pavir.9KG583700</td>
<td align="right">11325811</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">41598633</td>
<td align="left">Pavir.9KG583700.1</td>
<td align="left">Pavir.9KG583700.1.p</td>
<td align="left">PF07716</td>
<td align="left">PTHR22952,PTHR22952:SF198</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left"><a href="GO:0043565,GO:0006355,GO:0003700" class="uri">GO:0043565,GO:0006355,GO:0003700</a></td>
<td align="left">AT3G62420.1</td>
<td align="left">ATBZIP53,BZIP53</td>
<td align="left">basic region/leucine zipper motif 53</td>
<td align="left">LOC_Os03g19375.1</td>
<td align="left">NA</td>
<td align="left">expressed protein</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">Chr09N</td>
<td align="right">11315811</td>
<td align="right">11335811</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">4</td>
<td align="left">124936</td>
<td align="left">Pavir.9NG142800</td>
<td align="right">11325811</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">11329238</td>
<td align="right">11331352</td>
<td align="right">2115</td>
<td align="left">-</td>
<td align="right">41552177</td>
<td align="left">Pavir.9NG142800.1</td>
<td align="left">Pavir.9NG142800.1.p</td>
<td align="left">PF11250</td>
<td align="left">PTHR33155,PTHR33155:SF1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT5G22090.2</td>
<td align="left">NA</td>
<td align="left">Protein of unknown function (DUF3049)</td>
<td align="left">LOC_Os03g49830.1</td>
<td align="left">NA</td>
<td align="left">expressed protein</td>
<td align="right">-3427</td>
<td align="right">-5541</td>
<td align="right">3427</td>
</tr>
<tr class="odd">
<td align="left">Chr09N</td>
<td align="right">11315811</td>
<td align="right">11335811</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">4</td>
<td align="left">124937</td>
<td align="left">Pavir.9NG142800</td>
<td align="right">11325811</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">11329238</td>
<td align="right">11331352</td>
<td align="right">2115</td>
<td align="left">-</td>
<td align="right">41552177</td>
<td align="left">Pavir.9NG142800.1</td>
<td align="left">Pavir.9NG142800.1.p</td>
<td align="left">PF11250</td>
<td align="left">PTHR33155,PTHR33155:SF1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT5G22090.2</td>
<td align="left">NA</td>
<td align="left">Protein of unknown function (DUF3049)</td>
<td align="left">LOC_Os03g49830.1</td>
<td align="left">NA</td>
<td align="left">expressed protein</td>
<td align="right">-3427</td>
<td align="right">-5541</td>
<td align="right">3427</td>
</tr>
<tr class="even">
<td align="left">Chr09N</td>
<td align="right">11957533</td>
<td align="right">11977533</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">intron</td>
<td align="right">5</td>
<td align="left">111994</td>
<td align="left">Pavir.9KG622000</td>
<td align="right">11967533</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">NA</td>
<td align="right">41599909</td>
<td align="left">Pavir.9KG622000.1</td>
<td align="left">Pavir.9KG622000.1.p</td>
<td align="left">PF01588,PF09334</td>
<td align="left">PTHR11946,PTHR11946:SF78</td>
<td align="left">NA</td>
<td align="left">6.1.1.10</td>
<td align="left">K01874</td>
<td align="left"><a href="GO:0000049,GO:0006418,GO:0005524,GO:0004812,GO:0000166" class="uri">GO:0000049,GO:0006418,GO:0005524,GO:0004812,GO:0000166</a></td>
<td align="left">AT4G13780.1</td>
<td align="left">NA</td>
<td align="left">methionine–tRNA ligase, putative / methionyl-tRNA synthetase, putative / MetRS, putative</td>
<td align="left">LOC_Os06g31210.1</td>
<td align="left">NA</td>
<td align="left">methionyl-tRNA synthetase, putative, expressed</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">Chr09N</td>
<td align="right">11957533</td>
<td align="right">11977533</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">5</td>
<td align="left">119578</td>
<td align="left">Pavir.9NG147300</td>
<td align="right">11967533</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">11964735</td>
<td align="right">11967667</td>
<td align="right">2933</td>
<td align="left">+</td>
<td align="right">41554068</td>
<td align="left">Pavir.9NG147300.1</td>
<td align="left">Pavir.9NG147300.1.p</td>
<td align="left">PF00227</td>
<td align="left">PTHR11599,PTHR11599:SF6</td>
<td align="left">KOG0177</td>
<td align="left">3.4.25.1</td>
<td align="left">K02734</td>
<td align="left"><a href="GO:0051603,GO:0005839,GO:0004298" class="uri">GO:0051603,GO:0005839,GO:0004298</a></td>
<td align="left">AT3G22630.1</td>
<td align="left">PBD1,PRCGB</td>
<td align="left">20S proteasome beta subunit D1</td>
<td align="left">LOC_Os03g48930.1</td>
<td align="left">NA</td>
<td align="left">peptidase, T1 family, putative, expressed</td>
<td align="right">2798</td>
<td align="right">-134</td>
<td align="right">134</td>
</tr>
<tr class="even">
<td align="left">Chr09N</td>
<td align="right">11957533</td>
<td align="right">11977533</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">5</td>
<td align="left">124997</td>
<td align="left">Pavir.9NG147373</td>
<td align="right">11967533</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">11967554</td>
<td align="right">11968429</td>
<td align="right">876</td>
<td align="left">-</td>
<td align="right">41561168</td>
<td align="left">Pavir.9NG147373.1</td>
<td align="left">Pavir.9NG147373.1.p</td>
<td align="left">PF14368</td>
<td align="left">PTHR33044,PTHR33044:SF14</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT2G48130.1</td>
<td align="left">NA</td>
<td align="left">Bifunctional inhibitor/lipid-transfer protein/seed storage 2S albumin superfamily protein</td>
<td align="left">LOC_Os07g07860.1</td>
<td align="left">NA</td>
<td align="left">LTPL76 - Protease inhibitor/seed storage/LTP family protein precursor, expressed</td>
<td align="right">-21</td>
<td align="right">-896</td>
<td align="right">21</td>
</tr>
<tr class="odd">
<td align="left">Chr09N</td>
<td align="right">11957533</td>
<td align="right">11977533</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">5</td>
<td align="left">124998</td>
<td align="left">Pavir.9NG147500</td>
<td align="right">11967533</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">11968815</td>
<td align="right">11969741</td>
<td align="right">927</td>
<td align="left">-</td>
<td align="right">41554575</td>
<td align="left">Pavir.9NG147500.1</td>
<td align="left">Pavir.9NG147500.1.p</td>
<td align="left">PF14368</td>
<td align="left">PTHR33044,PTHR33044:SF14</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT3G22600.1</td>
<td align="left">NA</td>
<td align="left">Bifunctional inhibitor/lipid-transfer protein/seed storage 2S albumin superfamily protein</td>
<td align="left">LOC_Os03g57970.1</td>
<td align="left">NA</td>
<td align="left">LTPL73 - Protease inhibitor/seed storage/LTP family protein precursor, expressed</td>
<td align="right">-1282</td>
<td align="right">-2208</td>
<td align="right">1282</td>
</tr>
<tr class="even">
<td align="left">Chr09N</td>
<td align="right">11957533</td>
<td align="right">11977533</td>
<td align="right">20001</td>
<td align="left">+</td>
<td align="left">spliceSite</td>
<td align="right">5</td>
<td align="left">124999</td>
<td align="left">Pavir.9NG147573</td>
<td align="right">11967533</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1.1e+07</td>
<td align="left">Chr09N_1.1e+07</td>
<td align="right">11970419</td>
<td align="right">11974568</td>
<td align="right">4150</td>
<td align="left">-</td>
<td align="right">41553090</td>
<td align="left">Pavir.9NG147573.1</td>
<td align="left">Pavir.9NG147573.1.p</td>
<td align="left">PF07466</td>
<td align="left">PTHR33975,PTHR33975:SF2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">AT1G54520.1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">LOC_Os03g48920.1</td>
<td align="left">NA</td>
<td align="left">DUF1517 domain containing protein, putative, expressed</td>
<td align="right">-2886</td>
<td align="right">-7035</td>
<td align="right">2886</td>
</tr>
</tbody>
</table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
