<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SNP file inputs and QC • switchgrassGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SNP file inputs and QC">
<meta property="og:description" content="switchgrassGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">switchgrassGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/annotations.html">Annotation tables for top SNPs</a>
    </li>
    <li>
      <a href="../articles/fbm.html">Using file-backed matrices - FBM</a>
    </li>
    <li>
      <a href="../articles/gwas.html">Genome-Wide Association</a>
    </li>
    <li>
      <a href="../articles/inputs.html">SNP file inputs and QC</a>
    </li>
    <li>
      <a href="../articles/mash.html">Multivariate Adaptive Shrinkage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="inputs_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SNP file inputs and QC</h1>
            
      
      
      <div class="hidden name"><code>inputs.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="i-have-downloaded-a-bed-file-of-switchgrass-snp-data--how-do-i-use-this-with-the-switchgrassgwas-r-package">I have downloaded a bed file of switchgrass SNP data. How do I use this with the switchgrassGWAS R package?<a class="anchor" aria-label="anchor" href="#i-have-downloaded-a-bed-file-of-switchgrass-snp-data--how-do-i-use-this-with-the-switchgrassgwas-r-package"></a>
</h2>
<p>First, check that you have downloaded three files with identical file names and three different file type extensions: “.bed”, “.bim”, and “.fam”. These three files need to be in the same folder for this to work.</p>
<p>You can then do this within R with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install bigsnpr </span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"bigsnpr"</span><span class="op">)</span>

<span class="co"># Load bigsnpr package into your R environment</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>  

<span class="co"># Create the "bigSNP" object from the bed file</span>
<span class="va">rdsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="st">"path/to/your/bedfile.bed"</span>, backingfile <span class="op">=</span> <span class="st">"path/to/new/bigSNPfile"</span><span class="op">)</span>

<span class="co"># Attach the new "bigSNP" object to the R session.</span>
<span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"path/to/new/bigSNPfile"</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># You're ready to rock!</span></code></pre></div>
<p>In the <code>snp_readBed</code> function, you will need to replace “path/to/your/bedfile.bed” with the path to your .bed file, and you should replace “path/to/new/bigSNPfile” with your desired output file name and path to that file. Don’t add any extension to specify the file type; that will be done by bigsnpr.</p>
<p>You can also try doing this with the example file that comes with the switchgrassGWAS R package; see the commands to do that <a href="https://alice-macqueen.github.io/switchgrassGWAS/articles/gwas.html" class="external-link">here</a></p>
</div>
<div class="section level2">
<h2 id="i-have-a-vcf-file-of-switchgrass-snp-data--how-do-i-use-this-with-the-switchgrassgwas-r-package">I have a vcf file of switchgrass SNP data. How do I use this with the switchgrassGWAS R package?<a class="anchor" aria-label="anchor" href="#i-have-a-vcf-file-of-switchgrass-snp-data--how-do-i-use-this-with-the-switchgrassgwas-r-package"></a>
</h2>
<p>You will need to convert this vcf (or vcf.gz) file to a bed file, then follow the instructions in the previous section.</p>
<p>One way to do this is using the command line program <a href="http://zzz.bwh.harvard.edu/plink/index.shtml" class="external-link">plink</a>.</p>
<p>On the command line:</p>
<pre><code>plink --vcf path/to/your/vcf/or/vcf.gz --const-fid Pvirgatum --allow-extra-chr --make-bed --out outputfilename</code></pre>
<p>You will need to replace <code>path/to/your/vcf/or/vcf.gz</code> with the path to your vcf or vcf.gz file, or just the vcf or vcf.gz filename if you are running the command from the same directory as the file is in.</p>
<p>You will need to replace <code>outputfilename</code> with your desired output file name. Don’t add any extension to specify the file type; that will be done by plink.</p>
<p>Here’s an example command.</p>
<pre><code>plink --vcf Pvirgatum_V5_GWAS_758g_33M.vcf.gz --const-fid Pvirgatum --allow-extra-chr --make-bed --out Pvirgatum_V5_GWAS_758g_33M</code></pre>
<p>Alternatively, if your file is in vcf format (but not vcf.gz format), you can do this from within R using the following commands:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="va">prefix</span> <span class="op">&lt;-</span> <span class="st">"Pvirgatum_V5_GWAS_prefix"</span>  <span class="co">#  specify your file prefix (with no extension)</span>

<span class="co"># Download plink to use it from within R</span>
<span class="va">plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/download_plink.html" class="external-link">download_plink</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Create a bedfile from the vcf file (with some switchgrass-specific extra options)</span>
<span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_plinkQC.html" class="external-link">snp_plinkQC</a></span><span class="op">(</span>plink.path <span class="op">=</span> <span class="va">plink</span>, prefix.in <span class="op">=</span> <span class="va">prefix</span>,
                       file.type <span class="op">=</span> <span class="st">"--vcf"</span>, 
                       extra.options <span class="op">=</span> <span class="st">"--allow-no-sex --allow-extra-chr --chr Chr01K Chr01N Chr02K Chr02N Chr03K Chr03N Chr04K Chr04N Chr05K Chr05N Chr06K Chr06N Chr07K Chr07N Chr08K Chr08N Chr09K Chr09N"</span><span class="op">)</span>

<span class="co"># Create the "bigSNP" object from the bed file</span>
<span class="va">rdsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="va">prefix</span><span class="op">)</span>
         </code></pre></div>
</div>
<div class="section level2">
<h2 id="id-like-to-do-some-quality-control-qc-on-my-snps-before-proceeding--what-do-you-recommend">I’d like to do some quality control (QC) on my SNPs before proceeding. What do you recommend?<a class="anchor" aria-label="anchor" href="#id-like-to-do-some-quality-control-qc-on-my-snps-before-proceeding--what-do-you-recommend"></a>
</h2>
<p>Great idea! I recommend subsetting your SNP file to just the individuals you have phenotypes for, then removing SNPs that don’t meet specific QC thresholds. I’d recommend removing SNPs that have a minor allele count of less than 25, that are missing at a frequency of &gt;10%, and/or that are not in Hardy-Weinberg equilibrium.</p>
<div class="section level3">
<h3 id="why-remove-individuals">Why remove individuals?<a class="anchor" aria-label="anchor" href="#why-remove-individuals"></a>
</h3>
<p>If you have lots of phenotypes, there’s no need remove individuals on a phenotype-by-phenotype basis. However, if you have many phenotypes but no phenotypes for a particular individual, you gain no statistical signal by including that individual in your genome-wide association. If you include enough individuals with no phenotypes, that may introduce statistical artifacts and noise in SNPs that have low minor allele counts or high missingness in your phenotyped individual set (see next section).</p>
</div>
<div class="section level3">
<h3 id="why-remove-snps">Why remove SNPs?<a class="anchor" aria-label="anchor" href="#why-remove-snps"></a>
</h3>
<p>These SNP datasets are very large, and a single genome-wide association in switchgrass will involve millions of tests, with an unknown ratio of statistical signal to statistical noise. SNPs with low minor allele counts or high missingness - either genotypically or phenotypically - are more likely to be statistical ‘bad actors’ with high levels of statistical noise; removing them often dramatically improves the ratio of statistical signal to noise in the resulting association set.</p>
<p>You can do this QC within R with:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>

<span class="co"># Load your current bigSNP file into the R session.</span>
<span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="st">"path/to/your/snpfile.rds"</span><span class="op">)</span>

<span class="co"># Load your phenotypes into the R session.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pvdiv_phenotypes</span><span class="op">)</span> <span class="co"># load your own phenotypes, of course</span>

<span class="co"># Filter to individuals that you have at least one phenotype for</span>
<span class="co"># In this example, we only keep non-NA individuals for the bio5 phenotype</span>
<span class="va">my_ind</span> <span class="op">&lt;-</span> <span class="va">pvdiv_phenotypes</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">bio5</span><span class="op">)</span><span class="op">)</span> 

<span class="co"># Make numeric vector to select the individuals to keep using bigsnpr</span>
<span class="va">ind_selector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">snp</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">sample.ID</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">my_ind</span><span class="op">$</span><span class="va">PLANT_ID</span><span class="op">)</span>
 <span class="co"># Choose subset name</span>
<span class="va">subset_name</span> <span class="op">&lt;-</span> <span class="st">"Pvirgatum_V5_GWAS_phenotyped_subset"</span>

<span class="co"># Subset SNP files to phenotyped individuals</span>
<span class="va">snpfile_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">snp</span>, ind.row <span class="op">=</span> <span class="va">ind_selector</span><span class="op">)</span>
<span class="co"># Attach new SNP subset file for QC</span>
<span class="va">snp_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="va">snpfile_subset</span><span class="op">)</span>
<span class="co"># Save SNP subset file as a bed file for QC</span>
<span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_writeBed.html" class="external-link">snp_writeBed</a></span><span class="op">(</span><span class="va">snp_subset</span>, bedfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subset_name</span>, <span class="st">".bed"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Choose minor allele count cutoff. If your dataset is small, use a minor allele frequency cutoff of 5% instead.</span>
<span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">25</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">selector</span><span class="op">)</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># what should the MAF be?</span>
<span class="kw">if</span><span class="op">(</span><span class="va">mac</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">{</span>
  <span class="va">mac</span> <span class="op">&lt;-</span> <span class="fl">0.05</span>
<span class="op">}</span>
<span class="co"># Download plink to use it from within R</span>
<span class="va">plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/download_plink.html" class="external-link">download_plink</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Do QC within R using plink, with some switchgrass-specific settings</span>
<span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_plinkQC.html" class="external-link">snp_plinkQC</a></span><span class="op">(</span><span class="va">plink</span>, prefix.in <span class="op">=</span> <span class="va">subset_name</span>,
                       maf <span class="op">=</span> <span class="va">mac</span>, geno <span class="op">=</span> <span class="fl">0.1</span>, hwe <span class="op">=</span> <span class="fl">1e-50</span>, 
                       prefix.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subset_name</span>, <span class="st">"_maf"</span>, <span class="va">mac</span><span class="op">)</span>,
                       extra.options <span class="op">=</span> <span class="st">"--allow-no-sex --allow-extra-chr --chr Chr01K Chr01N Chr02K Chr02N Chr03K Chr03N Chr04K Chr04N Chr05K Chr05N Chr06K Chr06N Chr07K Chr07N Chr08K Chr08N Chr09K Chr09N"</span><span class="op">)</span>

<span class="co"># Create a new "bigSNP" object from the bed file</span>
<span class="va">rdsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed</a></span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="st">"path/to/new/bigSNPfile"</span><span class="op">)</span>

<span class="co"># You're done, but may want to remove the intermediate bigSNP and bed files from the directory where you were working.</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alice MacQueen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
